### DISCUSSION TITLE
Crystal plasticity Kalidini model error says you need to provide deformation gradient

### DISCUSSION THREAD

**Discussion by mahyarmalek** [idaholab/moose#20928] [Q&A Modules: Solid mechanics] [answered]
Hello, 

I'm trying to use the "ComputeMultipleCrytalPlasticity" model for Kalidini type and im getting these errors: 

```
Material property 'deformation_gradient', requested by 'stress_ice' is not defined on block 0
Material property 'deformation_gradient', requested by 'stress_ice_face' is not defined on block 0
Material property 'deformation_gradient', requested by 'stress_ice_neighbor' is not defined on block 0
```

Note that I have added "ComputeFiniteStrain" method to provide this and also I'm using this stress inside my own kernel, therefore I cannot use the tensor mechanics action here. I think the ComputeFiniteStrain method is not providing those info to the plasticity model. here's the input file for my material and kernel (I just put  those parts which might be usefull to read since it's a hige file) 

```
[Materials]
 [./elasticity_tensor_ice]
    type = ComputeElasticityTensorConstantRotationCP
    C_ijkl = '1.684e3 1.214e3 1.214e3 1.684e3 1.214e3 1.684e3 0.754e3 0.754e3 0.754e3'
    fill_method = symmetric9
    block = 0
    base_name = ice
  [../]
  [./stress_ice]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'trial_xtalpl'
    tan_mod_type = exact
    maximum_substep_iteration = 200
    use_line_search = true
    min_line_search_step_size = 0.01
    block = 0
    base_name = ice
  [../]
  [./trial_xtalpl]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    resistance_tol = 0.01
  [../]
    [./strain_ice]
    type = ComputeFiniteStrain
    displacements = 'u_f_x u_f_y'
    base_name = ice
    block = 0
   #volumetric_locking_correction=true
  [../]
  
```

Best, 

Mahyar

**Comment by GiudGiud** [comment]
Hello

Have you tried using the TensorMechanicsAction instead of adding the computeFiniteStrain material?

```
[Modules/TensorMechanics/Master/all]
  strain = FINITE
  add_variables = true
[]
```

Guillaume

**Reply by mahyarmalek** [reply]
I mentioned in the quation that i cannot use the action since my kernel is coupled with another variables. Is there any way to use the TensorMechanicsAction while my equation looks like this====> (div(sigma*phi) )= 0 in which sigma is stress and phi is phasefield variable coupled to this equation?

**Comment by jiangwen84** [comment, marked as answer]
are you using any interface kernel, e.g. cohesive zone materials? I asked this because I saw "neighbor" material is being asked.

If you have only one block, you do not need to use "base_name". Leave it as blank or default.

**Reply by mahyarmalek** [reply]
thank you for answering, I'm not using any cohesive zone material but my linear momentum equation is not a simple stress divergence. it is coupled with a phasefield variable which makes it impossible to use tensor mechanics master action. Also im using three different stresses with three different base_names in this equations (since im modeling a multiphase material). For the ice part I wanted to use the crystal plasticity model. So I imagine if ComputeFiniteStress cannot feed my ice_stress the proper deformation gradient then I will not be able to use this model.

**Reply by jiangwen84** [reply]
If you could attach your input file here, we can take a close look.

**Reply by mahyarmalek** [reply]
Sure here is the input file:

[main.txt](https://github.com/idaholab/moose/files/8599983/main.txt)

**Reply by jiangwen84** [reply]
I cannot directly run your input file because there are a few customized objects.

I took a close look at the codes. It seems that we need to make a change in line 92 and 93 in `ComputeMultipleCrystalPlasticityStress` by adding a `base_name`. 

```
_deformation_gradient(getMaterialProperty<RankTwoTensor>(_base_name + "deformation_gradient")), _deformation_gradient_old(getMaterialPropertyOld<RankTwoTensor>(_base_name +"deformation_gradient")), 
```

Could you make the change on your local branch and see if it fixes your problem?

**Reply by mahyarmalek** [reply]
Hello, 

Sorry for getting back to you late, I have noticed this problem and compiled the codes again with the base_name and it started working. It's better to take a look at all the plasticity classes since I observed this problem in other crystal plasticity codes too. However I could not make the code converge and it stops due to convergence inside the crystal plasticty model. I would blame it on complexity of my model and not being able to use the exact jacobians like AD provides. 

Best,

**Reply by jiangwen84** [reply]
@mahyarmalek Could you submit an ISSUE on GitHub? We would like to fix it.

**Reply by GiudGiud** [reply]
@jiangwen84 any update on this fix?
I reckon we can create the issue ourselves if needed

**Reply by jiangwen84** [reply]
Looks like @mahyarmalek found similar issues in other CP codes. I would suggest him add a MOOSE issue to list all issues.

**Reply by GiudGiud** [reply]
@mahyarmalek would be great to create an issue with what you observed. This is how we track development needs and limitations so it would be really valuable

**Reply by mahyarmalek** [reply]
Sorry for getting back to you late, I have created an issue, however other than ComputeMultipleCrystalPlasticity.C which I remember needed modification I can't recall where else did I see the problem but anyway I made the issue and for my case that was enough to work.

Thank you for your guidance! 

Mahyar

**Reply by GiudGiud** [reply]
great, thank you!

**Reply by sapitts** [reply]
Hi @mahyarmalek,
We've addressed the issue with the StressUpdate-based crystal plasticity code that you identified in PR #21169, and these changes are now merged into MOOSE. Thank you for creating the issue!
